connectors/sources/gcs.py

```diff
@@ -0,0 +1,346 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""Google Cloud Storage source module responsible to fetch documents from Google Cloud Storage.
+"""
+import asyncio
+import json
+import os
+import urllib.parse
+from functools import partial
+
+from aiofiles.tempfile import TemporaryFile
+from aiogoogle import Aiogoogle
+from aiogoogle.auth.creds import ServiceAccountCreds
+
+from connectors.logger import logger
+from connectors.utils import get_base64_value, TIKA_SUPPORTED_FILETYPES
+from connectors.source import BaseDataSource
+
+CLOUD_STORAGE_READ_ONLY_SCOPE = "https://www.googleapis.com/auth/devstorage.read_only"
+CLOUD_STORAGE_BASE_URL = "https://console.cloud.google.com/storage/browser/_details/"
+API_NAME = "storage"
+API_VERSION = "v1"
+BLOB_ADAPTER = {
+    "_id": "id",
+    "component_count": "componentCount",
+    "content_encoding": "contentEncoding",
+    "content_language": "contentLanguage",
+    "created_at": "timeCreated",
+    "last_updated": "updated",
+    "metadata": "metadata",
+    "name": "name",
+    "size": "size",
+    "storage_class": "storageClass",
+    "_timestamp": "updated",
+    "type": "contentType",
+    "url": "selfLink",
+    "version": "generation",
+    "bucket_name": "bucket",
+}
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_WAIT_MULTIPLIER = 2
+DEFAULT_CONTENT_EXTRACTION = True
+DEFAULT_FILE_SIZE_LIMIT = 10485760
+STORAGE_EMULATOR_HOST = os.getenv(
+    key="STORAGE_EMULATOR_HOST", default=None
+)  # For end to end tests set this environment variable with target host.
+RUNNING_FTEST = (
+    "RUNNING_FTEST" in os.environ
+)  # Flag to check if a connector is run for ftest or not.
+DEFAULT_PEM_FILE = (
+    os.path.join(
+        os.path.dirname(__file__),
+        "tests/fixtures/gcs",
+        "service_account_dummy_cert.pem",
+    ),
+)[0]
```

Comment:
why `[0]` here ?

Reply:
oh I see what you did. it's a `(tuple,)` -- not sure why, let's just us a plain string.